Implement wxFSVolume for macOS

TcT2k opened the pull request:
Implement with `NSFileManager` and `NSURL` to provide basic functionality like on MSW. Icons are not implemented for macOS.

TcT2k committed:
Implement wxFSVolume for macOS

Implement with NSFileManager and NSURL to provide basic functionality
like on MSW. Icons are not implemented for macOS.

vadz committed:
Rebake makefiles after adding src/osx/volume.mm

vadz commented:
Looks good to me, thanks! I've pushed another commit to actually compile the new file, let's see how it goes.

If there are no problems, and no objections from @csomor, I'll merge this soon.

TcT2k commented:
> Looks good to me, thanks! I've pushed another commit to actually compile the new file, let's see how it goes.
> 

Thank you for the bakefile rebake, I had only build with CMake so far.

vadz commented:
Note: X11 build failure is unrelated (installing httpbin failed again, we really need to replace it with something more robust... I just can't deal with Python any more at that point).

vadz commented:
... but iOS one is an actual problem. I don't know if we can build this code under iOS or if we need to exclude it from build there?

TcT2k committed:
Disable wxFSVolume for iOS

The underlying API NSFileManager
mountedVolumeURLsIncludingResourceValuesForKeys is available on iOS
but always returns an empty list so wouldn't be useful.

TcT2k commented:
> ... but iOS one is an actual problem. I don't know if we can build this code under iOS or if we need to exclude it from build there?

I disabled it for now as the underlying API `mountedVolumeURLsIncludingResourceValuesForKeys` returns an empty list  on iOS anyway (as documented by apple).

csomor commented:
@TcT2k thanks a lot for the patch, wxGetAvailableDrives was using wxFileIconsTable, perhaps we could use this as well ?

csomor commented on src/osx/volume.mm:
```diff
@@ -0,0 +1,142 @@
+    return m_isOk;
+}
+
+wxFSVolumeKind wxFSVolumeBase::GetKind() const
+{
+    NSURL* url = [NSURL fileURLWithPath:wxCFStringRef(GetName()).AsNSString()];
+    auto values = [url resourceValuesForKeys:@[NSURLVolumeIsLocalKey] error:nil];
+
+    // Assume disk for local volumes
+    if ([(NSNumber*)[values objectForKey:NSURLVolumeIsLocalKey] boolValue])
+        return wxFS_VOL_DISK;
+    else
+        return wxFS_VOL_NETWORK;
+}
```
To make `wxGetAvailableDrives` work as before, this must at least return cdrom for read only drives

```
wxFSVolumeKind wxFSVolumeBase::GetKind() const
{
    NSURL* url = [NSURL fileURLWithPath:wxCFStringRef(GetName()).AsNSString()];
    auto values = [url resourceValuesForKeys:@[NSURLVolumeIsLocalKey, NSURLVolumeIsReadOnlyKey] error:nil];

    // Assume disk for local volumes
    if ([(NSNumber*)[values objectForKey:NSURLVolumeIsLocalKey] boolValue])
    {
        if ([(NSNumber*)[values objectForKey:NSURLVolumeIsReadOnlyKey] boolValue])
            return wxFS_VOL_CDROM;
        else
            return wxFS_VOL_DISK;
    }
    else
        return wxFS_VOL_NETWORK;
}

```

csomor commented:
> > ... but iOS one is an actual problem. I don't know if we can build this code under iOS or if we need to exclude it from build there?
> 
> I disabled it for now as the underlying API `mountedVolumeURLsIncludingResourceValuesForKeys` returns an empty list on iOS anyway (as documented by apple).

but we could just return '/', and we'd support the wxVolume API also on iOS, and if things change, we could just enhance ?

```
   ...
    if ( nativeVolumes == nil )
    {
        wxFSVolumeBase volume("/");
        int flags = volume.GetFlags();
        if ((flags & flagsSet) == flagsSet && !(flags & flagsUnset))
                volumePaths.push_back(volume.GetName());
    }
   ....
```

csomor committed:
as per comments, rudimentary iOS implementation and read_only support in GetKind

csomor commented:
@TcT2k Hi Tobias, are these changes fine with you ? I've just committed them to this PR to make things easier to discuss and observe, if you and @vadz are fine with them, then please go ahead and merge it.

TcT2k commented:
> @TcT2k Hi Tobias, are these changes fine with you ? I've just committed them to this PR to make things easier to discuss and observe, if you and @vadz are fine with them, then please go ahead and merge it.

One potential issue with using read only as optical disc detection would be NTFS volumes on macOS (which are mounted as read only), this could be detected by additionally checking the file system type, but I don't think it's worth the effort.

So overall it look good to me and could be merged.

csomor commented:
> One potential issue with using read only as optical disc detection would be NTFS volumes on macOS (which are mounted as read only), this could be detected by additionally checking the file system type, but I don't think it's worth the effort.

thanks for mentioning this, I'll keep it in mind, perhaps this is too much of backwards compatibility ...

vadz commented:
Sorry, not sure what's the conclusion here, finally: should this be merged or are more changes needed, @csomor?

csomor commented:
> Sorry, not sure what's the conclusion here, finally: should this be merged or are more changes needed, @csomor?

If you're fine with it, then we can merge it

vadz commented:
This is clearly an improvement and while it's not perfect, let's merge it already and maybe improve it more later. Thanks again, Tobias!

vadz commented:
Oops, forgot to mention this PR in bbca67df678097e5c127b0300e47ef3f99520318 which (squash) merged it.
